instance: polymorphize upvar closures/generators

davidtwco opened the pull request:
This PR modifies how instances are polymorphized so that closures and generators have any closures or generators captured within their upvars also polymorphized.

With the new symbol mangling, a fully polymorphised closure will produce the same symbol regardless of what it was instantiated with. However, when that polymorphised closure captures another closure as an upvar, then the type of that other closure in the upvar substitution wouldn't have been polymorphised. The other closure will still refer to the initial substitutions. Therefore, the polymorphised closure will end up hashing differently but producing the same symbol - triggering `assert_symbols_are_distinct` in MIR partitioning. The old mangling scheme had a hash at the end that meant this didn't happen (this would still have been an issue, we just didn't have a way to notice).

See [this Zulip discussion for further elaboration](https://rust-lang.zulipchat.com/#narrow/stream/216091-t-compiler.2Fwg-polymorphization/topic/symbol.20mangling.20v0.20.E2.9C.95.20polymorphisation/near/206152008).

r? @eddyb
cc @lcnr

davidtwco committed:
instance: polymorphize upvar closures/generators

This commit modifies how instances are polymorphized so that closures
and generators have any closures or generators captured within their
upvars also polymorphized - this avoids symbol clashes with the new
symbol mangling scheme.

Signed-off-by: David Wood <david@davidtw.co>

lcnr commented on src/librustc_middle/ty/instance.rs:
```diff
@@ -502,6 +483,92 @@ impl<'tcx> Instance<'tcx> {
+    // from the function is captured as an upvar and hasn't been polymorphized. In this case,
+    // the unpolymorphized upvar closure would result in a polymorphized closure producing
+    // multiple mono items (and eventually symbol clashes).
+    let upvars_ty = if tcx.is_closure(def_id) {
```
is there a reason you aren't doing this as part of `unused_generic_params`?

davidtwco replied:
I don't think that would work; `unused_generic_params` could inspect the upvar tuple (it doesn't currently) and then invoke itself on any closure/generator upvars but if it went on to change that upvar closure (like this function does) then that wouldn't change anything outside of the query - by changing this here, it changes what goes into the `MonoItem`.

lcnr replied:
I had a misconception on what we are actually doing here :+1:

lcnr commented on src/librustc_middle/ty/instance.rs:
```diff
@@ -502,6 +483,92 @@ impl<'tcx> Instance<'tcx> {
+                    tcx.mk_param_from_def(param),
+
+            // Otherwise, use the parameter as before.
+            _ => substs[param.index as usize],
```
can't we just always polymorphize here ?
```suggestion
            _ => {
                let polymorphized_ty = substs[param.index as usize].fold_with(
                        &mut PolymorphizationFolder { tcx });
                ty::GenericArg::from(polymorphized_ty)
            },
```

lcnr replied:
this would fix this while also optimizing things like

```rust
fn test<T>() {
    [1, 2, 3].iter().map(|_| ()).collect::<Vec<()>>();
}

fn main() {
    test::<String>();
    test::<Vec<String>>();
}
```

lcnr replied:
It would require us to also change that early exit here afaict.
```rust
    if unused.is_empty() {
        // Exit early if every parameter was used.
        return substs;
    }
```

davidtwco replied:
Applying this suggestion would also work (instead of adding another arm like the PR currently does). I avoided doing this because I figured it would be more expensive (traversing every substitution rather than just upvars). 

That said, I might be missing something, because I’m now sure how it would help with the example in your second comment (`test` and its closure would be polymorphised now and with your change, no?).

lcnr replied:
My concern is the following: https://rust.godbolt.org/z/fGeseM

We currently create 2 instances of `slice::Iter::map` here, while after this change we probably only create one

lcnr replied:
And I think this should actually happen somewhat frequently with iterator adapters in generic functions

davidtwco replied:
Fixed!

lcnr commented on src/test/ui/polymorphization/closure_in_upvar/fn.rs:
```diff
@@ -0,0 +1,29 @@
+    // Don't use `f` in `y`, but refer to `x` so that the closure substs contain a reference to
+    // `x` that will differ for each instantiation despite polymorphisation of the varying
+    // argument.
+    let y = || x(());
```
What happens if `y` considers `f` used and `x` doesn't?

davidtwco replied:
If `y` considers `f` used then it’ll produce two mono items with different symbols (because `f` varies) - with this PR, in both mono items, the upvar subst for `x` will be polymorphised. `x` would still produce a single mono item like now.

However, it occurs to me that we might still have a issue if `x` considers `f` used. In that case, this patch wouldn’t change the upvar substitutions in `y` for `x` and therefore we’d still produce two mono items that aren’t equal but produce the same symbol. Not sure how to deal with that case - don’t consider upvars in hashing and equality?

lcnr replied:
Can `y` consider `f` unused if it uses `x` which considers it used?

davidtwco replied:
> However, it occurs to me that we might still have a issue if x considers f used. In that case, this patch wouldn’t change the upvar substitutions in y for x and therefore we’d still produce two mono items that aren’t equal but produce the same symbol. Not sure how to deal with that case - don’t consider upvars in hashing and equality?

After discussion with @lcnr and testing locally, can confirm that we don't have an issue here.

> Can y consider f unused if it uses x which considers it used?

If we allowed this then I think we could run into issues with the symbol mangling again - because the upvar of `x` in `y` would differ depending on the instantiation of `f`, which would result in two mono items which have the same symbol, triggering `assert_symbols_are_distinct` with the new mangling scheme.

davidtwco replied:
I've added tests for the cases we've discussed here.

davidtwco committed:
instance: always polymorphize substs

By always polymorphizing substitutions, functions which take closures as
arguments (e.g. `impl Fn()`) can have fewer mono items when some of the
argument closures can be polymorphized.

Signed-off-by: David Wood <david@davidtw.co>

lcnr commented on src/librustc_middle/ty/instance.rs:
```diff
@@ -502,6 +483,62 @@ impl<'tcx> Instance<'tcx> {
+            match ty.kind {
+                ty::Closure(def_id, substs) => {
+                    let polymorphized_substs = polymorphize(self.tcx, def_id, substs);
+                    self.tcx.mk_closure(def_id, polymorphized_substs)
```
nit: it's slightly more performant to check `substs != polymorphized_substs` here to prevent unnecessary hashmap lookups, similar to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.reuse_or_mk_region

davidtwco replied:
Fixed!

lcnr commented on src/librustc_middle/ty/instance.rs:
```diff
@@ -502,6 +483,62 @@ impl<'tcx> Instance<'tcx> {
+                    let polymorphized_substs = polymorphize(self.tcx, def_id, substs);
+                    self.tcx.mk_generator(def_id, polymorphized_substs, movability)
+                }
+                _ => ty.super_fold_with(self),
```
This is not that related to this PR itself, but why aren't we doing the same for `FnDef` here?

davidtwco replied:
At least for the original mangling issue, I wasn't sure that it could happen with a `FnDef`.

lcnr replied:
this should allow us to optimize the following I think
```rust
fn unused<T>() -> u64 {
    42
}

fn test<T>() {
    std::iter::repeat(unused::<T>).take(10);
}

pub fn main() {
    test::<String>();
    test::<Vec<String>>();
}
```

lcnr replied:
This probably also has the mangling issue :thinking: not really sure about how mangling works for `FnDef` though

davidtwco replied:
Fixed!

lcnr reviewed (approved):
We may want to add a type flag like `MAY_POLYMORPHIZE` which checks if the ty is or contains a `Closure` or `Generator` so we don't have to always looking into the subst.

But apart from that this looks good to me, I don't think @eddyb also needs to look over this :thinking: 

r=me

davidtwco committed:
ty: add `MAY_POLYMORPHIZE` flag

This commit adds a `MAY_POLYMORPHIZE` which checks for closures and
generators so that polymorphization of substs does not need to traverse
every substs.

Signed-off-by: David Wood <david@davidtw.co>

davidtwco committed:
instance: avoid unnecessary `mk_` calls

This commit avoids unnecessary calls to `mk_closure` and `mk_generator`
by checking if the polymorphized substs match the original substs.

Signed-off-by: David Wood <david@davidtw.co>

davidtwco committed:
instance: polymorphize `FnDef` substs

This commit extends previous polymorphization of substs to polymorphize
`FnDef`.

Signed-off-by: David Wood <david@davidtw.co>

davidtwco commented:
I've addressed those comments, including adding the type flag.

r? @lcnr

lcnr commented on src/test/ui/polymorphization/closure_in_upvar/fnmut.rs:
```diff
@@ -0,0 +1,34 @@
+// compile-flags:-Zpolymorphize=on -Zsymbol-mangling-version=v0
+
+fn foo(f: impl Fn()) {
+    // Mutate a upvar from `x` so that it implements `FnMut`.
```
```suggestion
    // Mutate an upvar from `x` so that it implements `FnMut`.
```

davidtwco replied:
Fixed!

lcnr reviewed (approved):
r=me after the nit is fixed

davidtwco commented:
@bors r=lcnr

bors commented:
:pushpin: Commit ac50d61785ae5112b9b4e30a58cfcffe096b31ec has been approved by `lcnr`

<!-- @bors r=lcnr ac50d61785ae5112b9b4e30a58cfcffe096b31ec -->
<!-- homu: {"type":"Approved","sha":"ac50d61785ae5112b9b4e30a58cfcffe096b31ec","approver":"lcnr"} -->

bors commented:
:hourglass: Testing commit ac50d61785ae5112b9b4e30a58cfcffe096b31ec with merge 09f4c9f5082f78b0adcee83d3ab4337e000cd28e...
<!-- homu: {"type":"BuildStarted","head_sha":"ac50d61785ae5112b9b4e30a58cfcffe096b31ec","merge_sha":"09f4c9f5082f78b0adcee83d3ab4337e000cd28e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/959594906), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35854)
Approved by: lcnr
Pushing 09f4c9f5082f78b0adcee83d3ab4337e000cd28e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35854","checks-actions":"https://github.com/rust-lang-ci/rust/runs/959594906"},"merge_sha":"09f4c9f5082f78b0adcee83d3ab4337e000cd28e"} -->
